packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java

```diff
@@ -205,10 +327,203 @@ public List<ProjectDocument> searchProjectsForUser(
 		final Query onlyProjects = QueryBuilders.bool(b2 -> b2.must(sh -> sh.term(t -> t.field("type").value("project"))));
 
 		if (query != null) {
-			req.query(q -> q.bool(b -> b.must(permissionQuery).must(onlyProjects).must(query)));
-		} else {
-			req.query(q -> q.bool(b -> b.must(onlyProjects).must(permissionQuery)));
+			return QueryBuilders.bool(b -> b.must(permissionQuery).must(onlyProjects).must(query));
+		}
+
+		return QueryBuilders.bool(b -> b.must(onlyProjects).must(permissionQuery));
+	}
+
+	/**
+	 * Generate asset embeddings for a project
+	 *
+	 * @param projectId
+	 * @param asset
+	 * @return
+	 */
+	public Future<Void> generateAndUpsertProjectAssetEmbeddings(final UUID projectId, final TerariumAsset asset) {
+		if (!isRunningTestProfile() && !asset.getTemporary()) {
+			final String embeddingText = asset.getEmbeddingSourceText();
+			if (embeddingText == null) {
+				return null;
+			}
+
+			log.info("Dispatch async embedding generation for asset {}", asset.getId());
+
+			return CompletableFuture.runAsync(() -> {
+				new Thread(() -> {
+					try {
+						final TerariumAssetEmbeddings embeddings = embeddingService.generateEmbeddings(embeddingText);
+
+						final ProjectAssetEmbedding projectAssetEmbedding = new ProjectAssetEmbedding();
+						projectAssetEmbedding.setAssetId(asset.getId());
+						projectAssetEmbedding.setAssetType(AssetType.getAssetType(asset.getClass()));
+						projectAssetEmbedding.setEmbeddingId(embeddings.getEmbeddings().get(0).getEmbeddingId());
+						projectAssetEmbedding.setVector(embeddings.getEmbeddings().get(0).getVector());
+						projectAssetEmbedding.setSpan(embeddings.getEmbeddings().get(0).getSpan());
+
+						final ProjectDocument projectDoc = elasticService.get(
+							getAlias(),
+							projectId.toString(),
+							ProjectDocument.class
+						);
+
+						// update the embedding
+						if (projectDoc.getAssetEmbeddings() == null) {
+							projectDoc.setAssetEmbeddings(List.of(projectAssetEmbedding));
+						} else {
+							int index = -1;
+							for (int i = 0; i < projectDoc.getAssetEmbeddings().size(); i++) {
+								if (projectDoc.getAssetEmbeddings().get(i).getAssetId().equals(asset.getId())) {
+									index = i;
+									break;
+								}
+							}
+							if (index != -1) {
+								projectDoc.getAssetEmbeddings().set(index, projectAssetEmbedding);
+							} else {
+								projectDoc.getAssetEmbeddings().add(projectAssetEmbedding);
+							}
+						}
```

Comment:
Why not deleting the embeddings and adding the new generated ones?
I'm not sure I understand, maybe it's an ES logic to do it this way.

Reply:
The embeddings are part of a larger document. So we update them in place and overwrite the entire document.